### PR TITLE
Btb 41 create search result page

### PR DESCRIPTION
**Pull request by NSakuta**
# Issue
https://amiconsult.atlassian.net/browse/BTB-XX

## Changes

- 
- 

## Checklist

* [x] I tested the functionality on my local environment.
* [x] I made sure my code is easy to read and understand.
* [x] I provided sufficient automated tests for the code.
* [x] There a no conflicts from my-branch..main.

### CODE REVIEW

File: src/app/header/header.component.ts [hunk 1]
```diff
@@ -11,8 +12,11 @@ import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
   templateUrl: './header.component.html',
   styleUrls: ['./header.component.scss'],
 })
+
 export class HeaderComponent {
-  constructor(private router: Router) {}
+
+  constructor(private router: Router,
+    private queryData: QueryService) { }
```
Comment (jolle22) [inline src/app/header/header.component.ts:19]: queryService would be a better name for this variable.

File: src/app/product-detail/product-detail.component.ts [hunk 2]
```diff
@@ -1,8 +1,8 @@
-import {Component, OnDestroy, OnInit} from '@angular/core';
-import {ActivatedRoute} from '@angular/router';
-import {Subscription} from 'rxjs';
-import {ProductService} from '../api/products.service';
-import {IProduct} from '../Interface/products';
+import { Component, OnDestroy, OnInit } from '@angular/core';
+import { ActivatedRoute, Router } from '@angular/router';
```
Comment (jolle22) [inline src/app/product-detail/product-detail.component.ts:2]: Is the Router used in this component?

**Review by jolle22 — CHANGES_REQUESTED** [review]

**Review by jolle22 — CHANGES_REQUESTED** [review]

File: src/app/Interface/products.ts [hunk 3]
```diff
@@ -2,6 +2,10 @@ export interface IProduct {
   id: number,
   title: string,
   teaser: string,
-  icon: string
+  icon: string,
+  category: string,
+  image: string,
+  status: string,
+  releaseDate: any;
```
Comment (jolle22) [inline src/app/Interface/products.ts:9]: What's the exact type for "releaseDate" might it a be a string?

**Review by jolle22 — CHANGES_REQUESTED** [review]

File: src/app/api/products.service.ts [hunk 4]
```diff
@@ -34,6 +34,15 @@ export class ProductService {
       );
   }
 
+  getProductsByQuery(queryParams: string): Observable<any> {
+
+    return this.http.get<any>(this.productUrl + '/search/' + queryParams)
+      .pipe(
+        tap(data => console.log('queryParams: ', data)),
```
Comment (jolle22) [inline src/app/api/products.service.ts:41]: Please remove the console.log :)

**Review by jolle22 — CHANGES_REQUESTED** [review]

File: src/app/search/search.component.ts [hunk 5]
```diff
@@ -31,29 +59,115 @@ export class SearchComponent implements OnInit, OnDestroy {
...
+    this._selectedStatus = '';
+    this._selectedCategory = '';
+    this._selectedReleaseDate = '';
   }
 
   onSearch(): void {
-    this.router.navigate(['search', this._searchDetails]);
+    this.router.navigate(["search", this._searchDetails]);
+    this.query = this._searchDetails;
+    this.sub = this.productService.getProductsByQuery(this._searchDetails).subscribe({
```
Comment (jolle22) [inline src/app/search/search.component.ts:104]: When you have more than one .subscribe( in a component it makes sense to use 
takeUntil. For more info see:

https://stackoverflow.com/questions/58428791/angular-rxjs-observable-takeuntil-vs-unsubscribe-with-a-subscription

**Review by jolle22 — CHANGES_REQUESTED** [review]

File: src/app/header/header.component.ts (hunk 1 above)
Comment (NSakuta) [reply src/app/header/header.component.ts:20]: changed it to queryService

File: src/app/Interface/products.ts (hunk 3 above)
Comment (NSakuta) [reply src/app/Interface/products.ts:9]: yes, right! I've changed it to string

File: src/app/api/products.service.ts (hunk 4 above)
Comment (NSakuta) [reply src/app/api/products.service.ts:41]: done :)

File: src/app/search/search.component.ts (hunk 5 above)
Comment (NSakuta) [reply src/app/search/search.component.ts:104]: I have changed it to 'takeUntil' as you have shown me and it works good. thank you :)

File: src/app/search/search.component.ts [hunk 6]
```diff
@@ -11,49 +18,161 @@ import { ProductService } from '../api/products.service';
...
+
+  status = new FormControl();
 
-  constructor(
-    private route: ActivatedRoute,
+  _selectedCategory: string = '';
+
+  _selectedStatus: string = '';
+
+  _selectedReleaseDate: string = '';
```
[the remark this answers is not in the document]
Comment (NSakuta) [reply src/app/search/search.component.ts:47]: I made them private and added 'getter' function for them

File: src/app/search/search.component.scss [hunk 7]
```diff
@@ -1,20 +1,37 @@
...
+  max-width: 350px;
+  min-width: 270px;
+  height: 90%;
+  margin-bottom: 8px;
 }
-
 .card:hover {
-  background-color: rgba(128, 128, 128, 0.11);
+  background-color: $card-hover-bg;
+}
```
Comment (knicken) [inline src/app/search/search.component.scss:21]: The material guideline states, that hovering over cards on desktop elevates them:
https://material.io/components/cards#behavior

**Comment by knicken** [comment]
Visiting the search page via url (eg. http://localhost:4200/search/hack) will not show any results.

**Comment by knicken** [comment]
For the pagination of the search results it would be nice to use the new query params for the api (eg https://ac-trainee-store-api.herokuapp.com/products/search/hack?from=0&to=12)

File: src/app/Interface/products.ts [hunk 8]
```diff
@@ -2,6 +2,10 @@ export interface IProduct {
   id: number,
   title: string,
   teaser: string,
-  icon: string
+  icon: string,
+  category: string,
+  image: string,
+  status: string,
+  releaseDate: string;
 }
```
Comment (knicken) [inline src/app/Interface/products.ts:10]: Just an idea: We could split this interface into ITeaserProduct and IProduct

File: src/app/header/header.component.ts (hunk 1 above)
Comment (knicken) [reply src/app/header/header.component.ts:20]: Now the instance is called different of the class. Im not a fan of that

File: src/app/search/search.component.html [hunk 9]
```diff
@@ -1,25 +1,104 @@
...
+
+    <div *ngIf="!selectedCategory && !selectedStatus">
+      <mat-paginator [length]="this.allProductsBySearch.length"
+                      [pageSize]="12"
+                      aria-label="Select page"
+                      (page)="onPageChange($event, this.allProductsBySearch)">
+      </mat-paginator>
+    </div>
   </div>
 </div>
```
Comment (knicken) [inline src/app/search/search.component.html:104]: Still there is lots of bootstrap in this file

File: src/app/search/search.component.scss [hunk 10]
```diff
@@ -1,20 +1,37 @@
...
+mat-form-field::ng-deep.mat-form-field-underline {
+  width: 0;
+}
+.mat-form-field {
+  font-size: 16px;
+}
+mat-form-field::ng-deep.mat-form-field-infix {
+  width: 150px;
+  margin: 0 20px;
 }
```
Comment (knicken) [inline src/app/search/search.component.scss:37]: Imo this file could be less complex using the sass ampersand

File: src/app/search/search.component.html [hunk 11]
```diff
@@ -1,25 +1,104 @@
...
+      <div class="col d-flex justify-content-between align-items-start"
+            *ngFor="let product of visibleProducts"
+            (click)='toProductDetailPage(product.id)'
+            data-id="product.id">
+
+        <mat-card class="mat-card card">
+          <mat-card-header>
+            <mat-card-title>{{product.title}}</mat-card-title>
+          </mat-card-header>
+          <img mat-card-image [src]="product.image" alt="Photo of a Shiba Inu">
```
[the remark this answers is not in the document]
Comment (NSakuta) [reply src/app/search/search.component.html:77]: done

File: src/app/search/search.component.html (hunk 9 above)
Comment (NSakuta) [reply src/app/search/search.component.html:104]: I replaced bootstrap with Angular Material (Grid-list for layout)

File: src/app/search-page/search.component.ts [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+      this.visibleProducts = this.filteredProducts.slice(0, 12);
+    }
+  }
+
+  get selectedReleaseDate() {
+    return this._selectedReleaseDate;
+  }
+
+  onResize(event: any) {
+    this.breakPoint = (event.target.innerWidth < 1200) ? 2 : 3;
```
Comment (jolle22) [inline src/app/search-page/search.component.ts:110]: Why is this 1200 here and 1000 in ngOnInit? :)

**Review by jolle22 — CHANGES_REQUESTED** [review]

File: src/app/search-page/search.component.ts (hunk 12 above)
Comment (NSakuta) [reply src/app/search-page/search.component.ts]: :)) changed it to 1200 in ngOnInit

File: src/app/header/header.component.ts [hunk 13]
```diff
@@ -21,12 +23,17 @@ import { Router } from '@angular/router';
     ]),
   ],
 })
+
 export class HeaderComponent {
+
+  constructor(private router: Router,
+    private queryData: QueryService,
```
Comment (jolle22) [inline src/app/header/header.component.ts:30]: It should be 
private queryService: QueryService

**Review by jolle22 — CHANGES_REQUESTED** [review]

File: src/app/search-page/search-result.component/search-result.component.ts [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+    this.breakPoint = (event.target.innerWidth < 1200) ? 2 : 3;
+  }
+
+  toProductDetailPage(id: number): void {
+    this.router.navigate(["product", id]);
+  }
+
+  ngOnInit(): void {
+    this.breakPoint = (window.innerWidth < 1200) ? 2 : 3;
+  }
```
Comment (jolle22) [inline src/app/search-page/search-result.component/search-result.component.ts:29]: onResize and ngOnInit have almost the same content.
Could you merge this into one function? :)

**Review by jolle22 — CHANGES_REQUESTED** [review]

File: src/app/search-page/filter-component/filter.component.html [hunk 15]
```diff
@@ -0,0 +1,30 @@
+<div *ngIf="this.query">
```
Comment (jolle22) [inline src/app/search-page/filter-component/filter.component.html:1]: I think the code would be easier to understand if you would move the *ngIf="this.query" to the parent component.
Like this you could remove the query from this component.

**Review by jolle22 — CHANGES_REQUESTED** [review]

File: src/app/search-page/search.component.ts [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+  getStatusList() {
+    const unfilteredStatusList: string[] = [];
+
+    this.allProductsBySearch.map(item => {
+      unfilteredStatusList.push(item.status)
+    })
+    const uniqueStatusList = findUniqueItems(unfilteredStatusList);
+
+    this.statusList = uniqueStatusList;
+  }
```
Comment (jolle22) [inline src/app/search-page/search.component.ts:115]: These functions are almost identical.
When this is the case it makes sense to move the identical parts into a new function.

File: src/app/search-page/search.component.ts [hunk 17]
```diff
@@ -0,0 +1,226 @@
...
+      })
+
+    this.filterService.statusData$
+      .pipe(takeUntil(this.destroy$))
+      .subscribe(status => {
+        this.resetFilterValues();
+        this.selectedStatus = status;
+        this.filteredProducts = findItemsByStatus(this.selectedStatus, this.allProductsBySearch);
+        this.visibleProducts = this.filteredProducts.slice(0, 12);
+      })
```
Comment (jolle22) [inline src/app/search-page/search.component.ts:136]: These functions are almost identical.
When this is the case it makes sense to move the identical parts into a new function.

File: src/app/search-page/search.component.ts [hunk 18]
```diff
@@ -0,0 +1,226 @@
...
+                this.allProductsBySearch = data.results;
+                this.searchCount = data.count;
+                this.query = query;
+                this.visibleProducts = this.allProductsBySearch.slice(0, 12);
+                this.getCategoryList();
+                this.getStatusList();
+              }
+            })
+        }
+      });
```
Comment (jolle22) [inline src/app/search-page/search.component.ts:184]: These functions are almost identical.
When this is the case it makes sense to move the identical parts into a new function.

**Review by jolle22 — CHANGES_REQUESTED** [review]

File: src/app/search-page/search.component.ts (hunk 18 above)
Comment (jolle22) [reply src/app/search-page/search.component.ts:184]: Also this part might be similar to the onSearch function on top of this component.

**Review by jolle22 — CHANGES_REQUESTED** [review]

File: src/app/header/header.component.ts (hunk 13 above)
Comment (NSakuta) [reply src/app/header/header.component.ts:30]: done

File: src/app/search-page/filter-component/filter.component.html (hunk 15 above)
Comment (NSakuta) [reply src/app/search-page/filter-component/filter.component.html:1]: you are right! I moved it to the parent component

File: src/app/header/header.component.ts (hunk 13 above)
Comment (jolle22) [reply src/app/header/header.component.ts:30]: Did you push your changes? I still dont see the change. 😅

**Review by jolle22 — CHANGES_REQUESTED** [review]

File: src/app/search-page/search.component.ts (hunk 16 above)
Comment (NSakuta) [reply src/app/search-page/search.component.ts:115]: - [x] done: created one function getFilterList()

File: src/app/search-page/search.component.ts (hunk 18 above)
Comment (NSakuta) [reply src/app/search-page/search.component.ts:184]: - [x] combined into one function

File: src/app/search-page/search-result.component/search-result.component.ts (hunk 14 above)
Comment (NSakuta) [reply src/app/search-page/search-result.component/search-result.component.ts:29]: I've tried, but I do not understand how to merge it into one function, if the first function takes 'event' as a parameter and other function will be called by ngOnInit and has 'window' parameter.

**Comment by NSakuta** [comment]
> Visiting the search page via url (eg. http://localhost:4200/search/hack) will not show any results.

**Pull request closed without merging**